docs/design/field-references.md

```diff
@@ -116,6 +117,65 @@ spec:
           clusterServiceClassExternalName: user-provided-service
           clusterServicePlanExternalName: default
           parameters:
-            x: y
+            important: true
+            host: "{{a-binding:bindsecret#host}}"
             password: "{{a-binding:bindsecret#password}}"
 ```
+
+**Warning: it is not currently safe to have a truly secret field
+passed this way (cf `a-binding:bindsecret#password`) unless it's inserted
+into a secret, as it will be exposed in the parameters of the created object.**
+To do the example above correctly, we could instead construct a new secret object
+in the appropriate form for a `ServiceInstance` `parametersFrom` secret reference.
+In future, this [should be automatic](https://github.com/atlassian/smith/issues/233).
+
+## Early validation
+
+Having references inside an object or plugin means that the final
+shape of the object will only be known once the dependencies have
+been evaluated. However, because we would like to fail as quickly
+as possible if the user has entered invalid parameters, 'default'
+values can be specified (as placeholders) so that ServiceInstance
+objects and plugins can be evaluated against their json schemas.
+
+Modifying part of the example from the previous section:
+
+```yaml
+  - name: b
+    dependsOn:
+    - a-binding
+    spec:
+      object:
+        metadata:
+          name: ups-instance-2
+        apiVersion: servicecatalog.k8s.io/v1beta1
+        kind: ServiceInstance
+        spec:
+          clusterServiceClassExternalName: user-provided-service
+          clusterServicePlanExternalName: default
+          parameters:
+            important: true
+            host: "{{a-binding:bindsecret#host#\"http://example.com\"}}"
```

Comment:
I don't think it's that big an issue, because the defaults shouldn't be human generated.

Reply:
I've removed the escaping from the example. Hope that helps :P

Reply:
If we don't want to make Bundle human readable, I would vote for scrapping the mustache-like format completely, and replacing it with a fully expanded JSON object (not sure myself if kidding or serious). It would be much less ambiguous and confusing than what we have there.

Reply:
With the 'bundledefaults' object, it's not clear to me how you would cleanly specify this in the autowiring function? Is the idea that each autowiring function would output parallel sets of objects? This seems very heavy-weight. If the main problem is the reference format, we could devolve into something like:

{{smith}}:
  obj: o
  path: x.y
  default: z

Would that be better?

Reply:
something like
```
host: >
  {
    "type": "bind-secret-reference",
    "resourceName": "a-binding",
    "path": "host",
    "example": "http://example.com"
  }

Reply:
But then how do you know it's a reference without attempted to parse every json object? That's why the json object format suggested above does.

Reply:
But yes, example is a nicer word that default in this context, probably.